weave/wandb_interface/wandb_artifact_pusher.py

```diff
@@ -46,17 +49,25 @@ def write_artifact_to_wandb(
 
     if _lite_run is None:
         lite_run = InMemoryLazyLiteRun(
-            entity_name, project_name, group="weave_artifact_pushers", _hide_in_wb=True
+            entity_name,
+            project_name,
+            group="weave_artifact_pushers",
+            _hide_in_wb=True,
+            _always_create_project=_create_new_artifact,
+            _fetch_last_history_step=_create_new_artifact,
         )
     else:
         lite_run = _lite_run
 
-    # Ensure the artifact type exists
-    lite_run.i_api.create_artifact_type(
-        artifact_type_name=artifact_type_name,
-        entity_name=lite_run.run.entity,
-        project_name=lite_run.run.project,
-    )
+    lite_run.upsert_project()
```

Comment:
Why is this needed? I think this always runs anyway in your implementation

Reply:
I'm not a fan how `lite_run.run.entity` caused an upsert project query to happen. The fact that it was running in the background implicitly made it hard to detect. So I decided to refactor to make it explicit.

Reply:
Ok, i see. It strikes me as odd to have `should_create_project=_create_new_artifact` then call `lite_run.upsert_project()` immediately after. 

As a user of this API, how do I know when i need to call `upsert_project`? Why is that not covered with `should_create_project=_create_new_artifact`? I think you might be able to refactor this a bit more to collocate the related calls.

Reply:
I can understand why you'd expect `should_create_project` to take care of upserting the project. However, this boolean is only being used to create the project when creating the run. In this case, we need `upsert_project` to be called explicitly because the `lite_run.i_api.create_artifact_type` call below requires that the project always exists. If we don't explicitly call `upsert_project`, it actually fails some tests/code paths (i.e when publishing a brand new artifact to wandb to a project that doesn't exist yet). 

I can rename the boolean to `create_project_for_run` if you'd like, but that would leak internal implementation details to the caller.